Translate Chapter 14 to PTBR (#26)
JuliaIntro/JuliaIntroBR.jl: book/src/chap14.asciidoc

[[chap14]]
== Arquivos

> Este capítulo introduz a ideia de programas “persistentes” que mantêm dados permanentemente armazenados, e mostra como usar tipos diferentes de armazenamentos permanentes, como arquivos e banco de dados.

Reviewer: ```suggestion
Este capítulo introduz a ideia de programas “persistentes” que mantêm dados permanentemente armazenados e mostra como usar tipos diferentes de armazenamentos permanentes, como arquivos e banco de dados.
```

> === Persistência

Reviewer: Dar um "enter" para ajustar as linhas.
Linhas 5 e 6, vazias.

A maioria dos programas que nós vimos até agora são transitórios no sentido de que eles executam por um tempo e produzem alguma saída, mas ao finalizarem, seus dados desaparecem. Se você executar o programa novamente, ele inicia com um estado limpo.
(((transitório)))

Outros programas são _persistentes_: eles executam por um longo tempo (ou toda hora); eles mantém pelo menos um pouco dos dados em um armazenamento permanente (um disco rígido, por exemplo); e se eles forem encerrados e reiniciados, eles retornam aonde pararam.
(((persistente)))

> Exemplos de programas persistentes são sistemas operacionais, que são executados toda vez que um computador é ligado, e servidores web, que são executados toda hora, aguardando solicitações vindos da rede.

Reviewer: ```suggestion
Exemplos de programas persistentes são sistemas operacionais, que são executados toda vez que um computador é ligado e servidores web, que são executados toda hora, aguardando solicitações vindos da rede.
```

> Uma das maneiras mais simples que programas usam para manter seus dados é lendo e escrevendo _arquivos de texto_. Nós já vimos programas que lêm arquivos de texto; neste capítulo nós veremos programas que as escrevem.

Reviewer: ```suggestion
Uma das maneiras mais simples que os programas usam para manter seus dados é lendo e escrevendo _arquivos de texto_. Nós já vimos programas que leem arquivos de texto; neste capítulo nós veremos programas que as escrevem.
```

Reviewer: Ref.: https://duvidas.dicio.com.br/leem-ou-leem/#:~:text=Leem%20%C3%A9%20a%20forma%20conjugada,palavra%20l%C3%AAm%20tamb%C3%A9m%20est%C3%A1%20errada.

> Uma alternativa é armazenar o estado do programa em um banco de dados. Neste capítulo eu também irei apresentar como usar um banco de dados simples.

Reviewer: ```suggestion
Uma alternativa é armazenar o estado do programa em um banco de dados. Neste capítulo, iremos apresentar como usar um banco de dados simples.
```

[[reading_and_writing]]
=== Lendo e Escrevendo

> Um arquivo de texto é uma sequência de caracteres armazenado em um meio permanente como um disco rígido, ou memória flash. Nós vimos como abrir e ler um arquivo em <<reading_word_lists>>.

Reviewer: ```suggestion
Um arquivo de texto é uma sequência de caracteres armazenado em um meio permanente como um disco rígido ou memória flash. Nós vimos como abrir e ler um arquivo em <<reading_word_lists>>.
```

Para escrever em um arquivo, você tem que abri-lo com modo +"w"+ como segundo parâmetro:
(((open)))

[source,@julia-repl-test chap14]
----
julia> fout = open("saida.txt", "w")
IOStream(<file saida.txt>)
----

Se o arquivo já existe, abri-lo em modo escrita limpa os dados antigos e começa de novo, então tome cuidado! Se o arquivo não existe, um novo é criado. +open+ retorna um objeto arquivo e a função +write+ coloca dados dentro do arquivo.
(((write)))((("função", "Base", "write", see="write")))

[source,@julia-repl-test chap14]
----
julia> linha1 = "No meio do caminho tinha uma pedra,\n";

julia> write(fout, linha1)
36
----

O valor de retorno é o número de caracteres que foram escritos. O arquivo objeto acompanha aonde ele está, então se você chamar +write+ novamente, ela adiciona novos dados ao final do arquivo.

[source,@julia-repl-test chap14]
----
julia> linha2 = "tinha uma pedra no meio do caminho.\n";

julia> write(fout, linha2)
36
----

> Quando você acabar de escrever no arquivo, ele deve ser fechado.

Reviewer: ```suggestion
Quando você terminar de escrever no arquivo, ele deve ser fechado.
```

----

Se você não fechar o arquivo, ele é fechado quando o programa encerra.

=== Formatação

> O argumeto da função write precisa ser uma string, então se nós quisermos colocar outros valores no arquivo, nós temos que convertê-los para string. A maneira mais fácil de fazer isso é com +string+ ou interpolação de strings:

Reviewer: ```suggestion
O argumeto da função write precisa ser uma string, então se quisermos colocar outros valores no arquivo, temos que convertê-los para string. A maneira mais fácil de fazer isso é usando +string+ ou interpolação de strings:
```

(((string)))(((interpolação de strings)))

[source,@julia-repl-test chap14]
----
julia> fout = open("saida.txt", "w")
IOStream(<file saida.txt>)
julia> write(fout, string(150))
3
----

Uma alternativa é usar a familia de funções +print(ln)+.
(((print)))(((println)))

[source,@julia-repl-test chap14]
----
julia> camelos = 42
42

> julia> println(fout, "Eu vi $camels camelos.")

Reviewer: ```suggestion
julia> println(fout, "Eu vi $camelos camelos.")
```

Reviewer: $camels é traduzido?

----

[TIP]
====

> Uma alternativa mais poderosa é o macro +@printf@+ que imprime usando uma string de especificação de formato de estilo C, você pode ler mais a respeito em https://docs.julialang.org/en/v1/stdlib/Printf/

Reviewer: ```suggestion
Uma alternativa mais poderosa é a macro +@printf@+ que imprime usando uma string de especificação de formato de estilo C, você pode ler mais a respeito em https://docs.julialang.org/en/v1/stdlib/Printf/
```

(((@printf)))((("macro", "Printf", "@printf", see="@printf")))(((formatação)))
====

> === Filenames e Paths

Reviewer: @joaoluisro - Acho que seria interessante ver se esse subtítulo deve ser traduzido. Podemos discutir na reunião. Vou traduzir aqui, e aí, você decide se adiciona ou não.

> === Filenames e Paths

Reviewer: ```suggestion
=== Nomes de Arquivos e Caminhos
```

Arquivos são organizados em _diretórios_ (também chamados de “pastas”). Todo programa em execução possui um “diretório atual”, que é o diretório padrão para a maioria das operações. Por exemplo, quando você abre um arquivo para leitura, o Julia a procura no diretório atual.
(((diretório)))((("pasta", see="diretório")))

A função +pwd+ retorna o nome do diretório atual:
(((pwd)))((("função", "Base", "pwd", see="pwd")))

[source,jlcon]
----
julia> cwd = pwd()
"/home/ben"
----

> +cwd+ significa “current working directory” (diretório atual funcional). O resultado neste exemplo é +home/ben+, que é o diretório home do usuário chamado +ben+.

Reviewer: ```suggestion
+cwd+ significa “current working directory” (diretório atual funcional). O resultado nesse exemplo é +home/ben+, que é o diretório home do usuário chamado +ben+.
```

> Uma string como +"/home/ben"+ que identifica um arquivo ou diretório é chamado de _path_ (caminho).

Reviewer: ```suggestion
Uma string como +"/home/ben"+ que identifica um arquivo ou diretório é chamado de _caminho_.
```

> (((path)))

Reviewer: ```suggestion
(((caminho)))
```

> Um nome de arquivo simples, como +memo.txt+ também é considerado um path, mas é um _path relativo_ porque se refere ao diretório atual. Se o diretório atual é +/home/ben+, o nome do arquivo +memo.txt+ se referiria à +/home/ben/memo.txt+.

Reviewer: ```suggestion
Um nome de arquivo simples, como +memo.txt+ também é considerado um caminho, mas é um _caminho relativo_ porque se refere ao diretório atual. Se o diretório atual é +/home/ben+, o nome do arquivo +memo.txt+ se referiria à +/home/ben/memo.txt+.
```

> (((path relativo)))((("path", "relativo", see="path relativo")))

Reviewer: ```suggestion
(((caminho relativo)))((("caminho", "relativo", see="caminho relativo")))
```

> Um path que começa com +/+ não depende do diretório atual; ele é chamado de _path absoluto_. Para achar um path absoluto para um arquivo, você pode usar +abspath+:

Reviewer: ```suggestion
Um caminho que começa com +/+ não depende do diretório atual; ele é chamado de _caminho absoluto_. Para achar um caminho absoluto para um arquivo, você pode usar +abspath+:
```

> (((path absoluto)))((("path", "absoluto", see="path absoluto")))(((abspath)))((("função", "Base", "abspath", see="abspath")))

Reviewer: ```suggestion
(((caminho absoluto)))((("caminho", "absoluto", see="caminho absoluto")))(((abspath)))((("função", "Base", "abspath", see="abspath")))
```

[source,jlcon]
----
julia> abspath("memo.txt")
"/home/ben/memo.txt"
----

> O Julia fornece outras funções para trabalhar com nomes de arquivos e paths. Por exemplo, +ispath+ verifica se um arquivo ou diretório existe:

Reviewer: ```suggestion
O Julia fornece outras funções para trabalhar com nomes de arquivos e caminhos. Por exemplo, +ispath+ verifica se um arquivo ou diretório existe:
```

----

Se você não fechar o arquivo, ele é fechado quando o programa encerra.

=== Formatação

> O argumeto da função write precisa ser uma string, então se quisermos colocar outros valores no arquivo, temos que convertê-los para string. A maneira mais fácil de fazer isso é usando +string+ ou interpolação de strings:

Reviewer: ```suggestion
O argumento da função write precisa ser uma string, então se quisermos colocar outros valores no arquivo, temos que convertê-los para string. A maneira mais fácil de fazer isso é usando +string+ ou interpolação de strings:
```

(((string)))(((interpolação de strings)))

[source,@julia-repl-test chap14]
----
julia> fout = open("saida.txt", "w")
IOStream(<file saida.txt>)
julia> write(fout, string(150))
3
----

Uma alternativa é usar a familia de funções +print(ln)+.
(((print)))(((println)))

[source,@julia-repl-test chap14]
----
julia> camelos = 42
42
julia> println(fout, "Eu vi $camelos camelos.")

----

[TIP]
====

> Uma alternativa mais poderosa é a macro +@printf@+ que imprime usando uma string de especificação de formato de estilo C, você pode ler mais a respeito em https://docs.julialang.org/en/v1/stdlib/Printf/

Reviewer 2: ```suggestion
Uma alternativa mais poderosa é a macro +@printf+ que imprime usando uma string de especificação de formato de estilo C, você pode ler mais a respeito em https://docs.julialang.org/en/v1/stdlib/Printf/
```

(((@printf)))((("macro", "Printf", "@printf", see="@printf")))(((formatação)))
====

=== Nomes de Arquivos e Caminhos

Arquivos são organizados em _diretórios_ (também chamados de “pastas”). Todo programa em execução possui um “diretório atual”, que é o diretório padrão para a maioria das operações. Por exemplo, quando você abre um arquivo para leitura, o Julia a procura no diretório atual.
(((diretório)))((("pasta", see="diretório")))

A função +pwd+ retorna o nome do diretório atual:
(((pwd)))((("função", "Base", "pwd", see="pwd")))

[source,jlcon]
----
julia> cwd = pwd()
"/home/ben"
----

> +cwd+ significa “current working directory” (diretório atual funcional). O resultado nesse exemplo é +home/ben+, que é o diretório home do usuário chamado +ben+.

Reviewer 2: ```suggestion
+cwd+ significa “current working directory” (diretório de trabalho atual). O resultado nesse exemplo é +home/ben+, que é o diretório home do usuário chamado +ben+.
```

> Uma string como +"/home/ben"+ que identifica um arquivo ou diretório é chamado de _caminho_.

Reviewer: ```suggestion
Uma string como +"/home/ben"+ que identifica um arquivo ou diretório é chamada de _caminho_.
```

(((caminho)))

Um nome de arquivo simples, como +memo.txt+ também é considerado um caminho, mas é um _caminho relativo_ porque se refere ao diretório atual. Se o diretório atual é +/home/ben+, o nome do arquivo +memo.txt+ se referiria à +/home/ben/memo.txt+.
(((caminho relativo)))((("caminho", "relativo", see="caminho relativo")))

> Um caminho que começa com +/+ não depende do diretório atual; ele é chamado de _caminho absoluto_. Para achar um caminho absoluto para um arquivo, você pode usar +abspath+:

Reviewer: ```suggestion
Um caminho que começa com +/+ não depende do diretório atual; ele é chamado de _caminho absoluto_. Para encontrar um caminho absoluto para um arquivo, você pode usar +abspath+:
```

true
----

De forma similar, +isfile+ verifica se é um arquivo.
(((isfile)))((("função", "Base", "isfile", see="isfile")))

> +readdir+ retorna um array de arquivos (e outros diretórios) no diretório dado:

Reviewer: ```suggestion
+readdir+ retorna uma lista de arquivos (e outros diretórios) no diretório dado:
```

(((readdir)))((("função, "Base", "readdir", see="readdir")))

[source,jlcon]
----
julia> readdir(cwd)
3-element Array{String,1}:
 "memo.txt"

>  "musica"

Reviewer: ```suggestion
 "música"
```

 "fotos"
----

> Para demonstrar estas funções, o exemplo a seguir “anda” por um diretório, exibe os nomes de todos os arquivos, e chama a si mesmo recursivamente em todos os diretórios.

Reviewer: ```suggestion
Para demonstrar estas funções, o exemplo a seguir “caminha” por um diretório, exibe os nomes de todos os arquivos e chama a si mesmo recursivamente em todos os diretórios.
```

Reviewer: Duas opções. Veja qual você acha que fique mais fluído.

> Para demonstrar estas funções, o exemplo a seguir “anda” por um diretório, exibe os nomes de todos os arquivos, e chama a si mesmo recursivamente em todos os diretórios.

Reviewer: ```suggestion
Para demonstrar estas funções, o exemplo a seguir “percorre” por um diretório, exibe os nomes de todos os arquivos e chama a si mesmo recursivamente em todos os diretórios.
```

> (((anda)))((("função", "definido pelo programador", "anda", see="anda")))

Reviewer: ```suggestion
(((caminha)))((("função", "definido pelo programador", "caminha", see="caminha")))
```

[source,@julia-setup chap14]
----

> function anda(nomedir)

Reviewer: ```suggestion
function caminha(nomedir)
```

    for nome in readdir(nomedir)
        path = joinpath(nomedir, nome)
        if isfile(path)
            println(path)
        else

>             anda(path)

Reviewer: Não consegui colocar aqui. Mas, talvez trocar o nome da função por "caminha". Veja se fica melhor...

        end
    end
end
----

> +joinpath+ recebe um diretório e um nome de arquivo e junta-os em um path completo.

Reviewer: ```suggestion
+joinpath+ recebe um diretório e um nome de arquivo e junta-os em um caminho completo.
```

(((joinpath)))((("função", "Base", "joinpath", see="joinpath")))

[TIP]
====

> O Julia fornece uma função chamada +walkdir+ (veja https://docs.julialang.org/en/v1/base/file/#Base.Filesystem.walkdir) que é similar a esta mas é mais versátil. Como um exercício, leia a documentação e a use para imprimir os nomes dos arquivos em um dado diretório e seus subdiretórios.

Reviewer: ```suggestion
O Julia fornece uma função chamada +walkdir+ (consulte https://docs.julialang.org/en/v1/base/file/#Base.Filesystem.walkdir) que é similar a esta mas é mais versátil. Como um exercício, leia a documentação e a use para imprimir os nomes dos arquivos em um dado diretório e seus subdiretórios.
```

(((walkdir)))((("função", "Base", "walkdir", see="walkdir")))
====

[[catching_exceptions]]
=== Capturando Exceções

Muitas coisas podem dar errado quando você tenta ler e escrever arquivos. Se você tentar abrir um arquivo que não existe, você recebe um +SystemError+:
(((SystemError)))((("error", "Base", "SystemError", see="SystemError")))

[source,@julia-repl-test]
----
julia> fin = open("arquivo_ruim")
ERROR: SystemError: opening file "arquivo_ruim": No such file or directory
----

Se você não tem permissão para acessar o arquivo:

[source,@julia-repl-test]
----
julia> fout = open("/etc/passwd", "w")
ERROR: SystemError: opening file "/etc/passwd": Permission denied
----

> Para evitar estes erros, você pode usar funções como +ispath+ e +isfile+, mas levaria muito tempo e código para verificar todas as possibilidades.

Reviewer: ```suggestion
Para evitar estes erros, você pode usar funções como +ispath+ e +isfile+, porém, tomaria muito tempo e código para verificar todas as possibilidades.
```

É mais fácil tentar de uma vez-e lidar com os problemas se eles ocorrerem-que é exatamente o que a declaração +try+ faz. A sintaxe é similar a uma declaração +if+:
(((declaração try)))((("declaração", "try", see="declaração try")))(((try)))((("palavra-chave", "try", see="try")))(((catch)))((("palavra-chave", "catch", see="catch")))(((end)))

[source,julia]
----
try
    fin = open("arquivo_ruim.txt")
catch exc
    println("Algo deu errado: $exc")
end
----

> O Julia inicia executando a cláusula +try+. Se tudo der certo, ele pula a cláusula +catch+ e segue adiante. Se uma exceção ocorre, ele pula fora da cláusula +try+ e executa a cláusula +catch+.

Reviewer: ```suggestion
O Julia inicia executando a cláusula +try+. Se tudo der certo, ele pula a cláusula +catch+ e segue adiante. Se ocorrer alguma exceção, ele pula fora da cláusula +try+ e executa a cláusula +catch+.
```

> Lidar com uma exceção com uma declaração +try+ é chamado de _capturar_ uma exceção. Neste exemplo, a cláusula exceto imprime uma mensagem de erro que não é muito útil. Em geral, capturar uma exceção nos da uma chance de consertar o problema, ou tentar novamente, ou pelo menos encerrar o programa graciosamente.

Reviewer: ```suggestion
Lidar com uma exceção com uma declaração +try+ é chamado de _capturar_ uma exceção. Neste exemplo, a cláusula exceto imprime uma mensagem de erro que não é muito útil. Em geral, capturar uma exceção nos da uma chance de consertar o problema, ou tentar novamente ou pelo menos encerrar o programa graciosamente.
```

(((capturar uma exceção)))

> Em código que realiza mudanças de estado ou usa recursos como arquivos, há geralmente trabalho de limpeza (como fechar arquivos) que precisa ser feito quando o código é encerrado. Exceções potencialmente complicam esta tarefa, já que elas podem causar a saída de um bloco de código antes dele atingir seu fim normal. A palavra-chave +finally+ fornece uma maneira de executar código quando um dado bloco de código sai, independente de como ele saiu:

Reviewer: ```suggestion
Em código que realiza mudanças de estado ou usa recursos como arquivos, geralmente há um trabalho de limpeza (como fechar arquivos) que precisa ser feito quando o código é encerrado. Exceções potencialmente complicam esta tarefa, já que elas podem causar a saída de um bloco de código antes dele atingir seu fim normal. A palavra-chave +finally+ fornece uma maneira de executar código quando um dado bloco de código sai, independente de como ele saiu:
```

true
----

De forma similar, +isfile+ verifica se é um arquivo.
(((isfile)))((("função", "Base", "isfile", see="isfile")))

+readdir+ retorna uma lista de arquivos (e outros diretórios) no diretório dado:
(((readdir)))((("função, "Base", "readdir", see="readdir")))

[source,jlcon]
----
julia> readdir(cwd)
3-element Array{String,1}:
 "memo.txt"
 "música"
 "fotos"
----

Para demonstrar estas funções, o exemplo a seguir “caminha” por um diretório, exibe os nomes de todos os arquivos e chama a si mesmo recursivamente em todos os diretórios.
(((caminha)))((("função", "definido pelo programador", "caminha", see="caminha")))

[source,@julia-setup chap14]
----
function caminha(nomedir)
    for nome in readdir(nomedir)
        path = joinpath(nomedir, nome)
        if isfile(path)
            println(path)
        else

>             anda(path)

Reviewer 2: ```suggestion
            caminha(path)
```

        end
    end
end
----

+joinpath+ recebe um diretório e um nome de arquivo e junta-os em um caminho completo.
(((joinpath)))((("função", "Base", "joinpath", see="joinpath")))

[TIP]
====

> O Julia fornece uma função chamada +walkdir+ (consulte https://docs.julialang.org/en/v1/base/file/#Base.Filesystem.walkdir) que é similar a esta mas é mais versátil. Como um exercício, leia a documentação e a use para imprimir os nomes dos arquivos em um dado diretório e seus subdiretórios.

Reviewer: ```suggestion
O Julia fornece uma função chamada +walkdir+ (consulte https://docs.julialang.org/en/v1/base/file/#Base.Filesystem.walkdir) que é similar a esta, porém, é mais versátil. Como um exercício, leia a documentação e a use para imprimir os nomes dos arquivos em um dado diretório e seus subdiretórios.
```

(((walkdir)))((("função", "Base", "walkdir", see="walkdir")))
====

[[catching_exceptions]]
=== Capturando Exceções

Muitas coisas podem dar errado quando você tenta ler e escrever arquivos. Se você tentar abrir um arquivo que não existe, você recebe um +SystemError+:
(((SystemError)))((("error", "Base", "SystemError", see="SystemError")))

[source,@julia-repl-test]
----
julia> fin = open("arquivo_ruim")
ERROR: SystemError: opening file "arquivo_ruim": No such file or directory
----

Se você não tem permissão para acessar o arquivo:

[source,@julia-repl-test]
----
julia> fout = open("/etc/passwd", "w")
ERROR: SystemError: opening file "/etc/passwd": Permission denied
----

Para evitar estes erros, você pode usar funções como +ispath+ e +isfile+, porém, tomaria muito tempo e código para verificar todas as possibilidades.

> É mais fácil tentar de uma vez-e lidar com os problemas se eles ocorrerem-que é exatamente o que a declaração +try+ faz. A sintaxe é similar a uma declaração +if+:

Reviewer: Travessão
```suggestion
É mais fácil tentar de uma vez—e lidar com os problemas se eles ocorrerem—que é exatamente o que a declaração +try+ faz. A sintaxe é similar a uma declaração +if+:
```

(((declaração try)))((("declaração", "try", see="declaração try")))(((try)))((("palavra-chave", "try", see="try")))(((catch)))((("palavra-chave", "catch", see="catch")))(((end)))

[source,julia]
----
try
    fin = open("arquivo_ruim.txt")
catch exc
    println("Algo deu errado: $exc")
end
----

O Julia inicia executando a cláusula +try+. Se tudo der certo, ele pula a cláusula +catch+ e segue adiante. Se ocorrer alguma exceção, ele pula fora da cláusula +try+ e executa a cláusula +catch+.

> Lidar com uma exceção com uma declaração +try+ é chamado de _capturar_ uma exceção. Neste exemplo, a cláusula exceto imprime uma mensagem de erro que não é muito útil. Em geral, capturar uma exceção nos da uma chance de consertar o problema, ou tentar novamente ou pelo menos encerrar o programa graciosamente.

Reviewer 2: ```suggestion
Lidar com uma exceção com uma declaração +try+ é chamado de _capturar_ uma exceção. Neste exemplo, a cláusula de exceção imprime uma mensagem de erro que não é muito útil. Em geral, capturar uma exceção nos da uma chance de consertar o problema, ou tentar novamente ou pelo menos encerrar o programa graciosamente.
```

end
----

A função +close+ sempre será executada.

[[banco de dados]]
=== Banco de Dados

Um _banco de dados_ é um arquivo que é organizado para guardar dados. Muitos bancos de dados são organizados como um dicionário no sentido de que eles mapeiam chave para valores. A maior diferença entre um banco de dados e um dicionário é de que o banco de dados está em disco (ou em armazenamento permanente), então ele persiste após o programa encerrar.
(((banco de dados)))

> O ThinkJulia fornece uma interface para +GDBM+ (GNU dbm) para criar e atualizar arquivos de bancos de dados. Como um exemplo, irei criar um banco de dados que contém legendas para arquivos de imagens.

Reviewer 2: ```suggestion
O JuliaIntroBR fornece uma interface para +GDBM+ (GNU dbm) para criar e atualizar arquivos de bancos de dados. Como um exemplo, irei criar um banco de dados que contém legendas para arquivos de imagens.
```

(((GDBM)))

Abrir um banco de dados é similar a abrir outros arquivos:

> (((DBM)))((("tipo", "ThinkJulia", "DBM", see="DBM")))

Reviewer 2: ```suggestion
(((DBM)))((("tipo", "JuliaIntroBR", "DBM", see="DBM")))
```

[source,@julia-repl-test chap14]
----
julia> using ThinkJulia

julia> db = DBM("legendas", "c")
DBM(<legendas>)
----

O modo "c" significa que o banco de dados deve ser criado se ele ainda não existe. O resultado é um objeto banco de dados que pode ser usado (para a maioria das operações) como um dicionário.

Quando você cria um novo item, +GDBM+ atualiza o arquivo banco de dados:
(((operador colchetes)))

[source,@julia-repl-test chap14]
----
julia> db["cleese.png"] = "Foto de John Cleese."
"Foto de John Cleese."
----

Quando você acessa um de seus itens, +GDBM+ lê o arquivo:

[source,@julia-repl-test chap14]
----
julia> db["cleese.png"]
"Foto de John Cleese."
----

> Se você faz outra atribuição para uma chava já existe, +GDBM+ substitui o valor antigo:

Reviewer: ```suggestion
Se você faz outra atribuição para uma chave já existente, +GDBM+ substitui o valor antigo:
```

(((atribuição)))

[source,@julia-repl-test chap14]
----

> julia> db["cleese.png"] = "Foto de John Cleese fazendo uma andada engraçada."

Reviewer: ```suggestion
julia> db["cleese.png"] = "Foto de John Cleese fazendo uma caminhada engraçada."
```

> "Foto de John Cleese fazendo uma andada engraçada."

Reviewer: ```suggestion
"Foto de John Cleese fazendo uma caminhada engraçada."
```

julia> db["cleese.png"]

> "Foto de John Cleese fazendo uma andada engraçada."

Reviewer: ```suggestion
"Foto de John Cleese fazendo uma caminhada engraçada."
```

----

=== Serialização

> A limitação do +GDBM+ é de que as chaves e valores precisam ser strings ou arrays de bytes. Se você tentar usar qualquer outro tipo, você recebe um erro.

Reviewer: Acredito que aqui esteja tratando de _array_ e não de _listas_. Dê um _double check_ nessa parte caso precise mudar.

----

=== Serialização

A limitação do +GDBM+ é de que as chaves e valores precisam ser strings ou listas de bytes. Se você tentar usar qualquer outro tipo, você recebe um erro.

> As funções +serialize+ e +deserialize+ podem ajudar. Elas traduzem quase todo tipo de objeto em um array de bytes (um iobuffer) adequado para armazenamento em um banco de dados, e em seguida traduz o array de bytes em objetos:

Reviewer: ```suggestion
As funções +serialize+ e +deserialize+ podem ajudar. Elas traduzem quase todo tipo de objeto em uma lista de bytes (um iobuffer) adequado para armazenamento em um banco de dados e em seguida traduz a lista de bytes em objetos:
```

[1, 2, 3]
----

> +serialize+ e +deserialize+ escrevem para e lêm de um objeto iobuffer que representa um fluxo I/O em memória. A função +take!+ busca os conteúdos do iobuffer como um array de bytes e reinicia o iobuffer para seu estado inicial.

Reviewer: ```suggestion
+serialize+ e +deserialize+ escrevem para e leem de um objeto iobuffer que representa um fluxo I/O em memória. A função +take!+ busca os conteúdos do iobuffer como uma lista de bytes e reinicia o iobuffer para seu estado inicial.
```

> +serialize+ e +deserialize+ escrevem para e lêm de um objeto iobuffer que representa um fluxo I/O em memória. A função +take!+ busca os conteúdos do iobuffer como um array de bytes e reinicia o iobuffer para seu estado inicial.

Reviewer: Caso não for alterar para "lista de bytes"...
```suggestion
+serialize+ e +deserialize+ escrevem para e leem de um objeto iobuffer que representa um fluxo I/O em memória. A função +take!+ busca os conteúdos do iobuffer como um array de bytes e reinicia o iobuffer para seu estado inicial.
```

> Apesar do objeto novo ter o mesmo valor do antigo, ele não é (em geral) o mesmo objeto:

Reviewer: ```suggestion
Apesar do novo objeto ter o mesmo valor do antigo, ele não é (em geral) o mesmo objeto:
```

false
----

Em outras palavras, serialization e em seguida deserialization tem o mesmo efeito de copiar o objeto.
(((copiar)))

Você pode usar isso para guardar coisas diferentes de strings em um banco de dados.

[TIP]
====
Na verdade, o armazenamento de coisas diferentes de string em um banco de dados é tão comum que ele foi encapsulado em um pacote chamado +JLD2+ (ver https://github.com/JuliaIO/JLD2.jl).
====

=== Objetos de Comando

> A maioria dos sistemas operacionais fornecem uma interface de linha de comando, também conhecida como uma _shell_. Shells geralmente fornecem comandos para navegar o sistema de arquivos e iniciar aplicações. Por exemplo, no Unix você pode mudar de diretório com o comando +cd+, exibir o conteúdo do diretório com +ls+, e iniciar o web browser digitando (por exemplo) +firefox+.

Reviewer: ```suggestion
A maioria dos sistemas operacionais fornecem uma interface de linha de comando, também conhecida como uma _shell_. Shells geralmente fornecem comandos para navegar o sistema de arquivos e iniciar aplicações. Por exemplo, no Unix você pode mudar de diretório com o comando +cd+, exibir o conteúdo do diretório com +ls+ e iniciar o web browser digitando (por exemplo) +firefox+.
```

(((shell)))(((cd)))

Qualquer programa que você inicia do shell também pode ser iniciado do Julia usando um _objeto de comando_:
(((objeto de comando)))(((backticks)))((("``", see="backticks")))

[source,@julia-repl-test chap14]
----
julia> cmd = `echo olá`
`echo olá`
----

Acentos graves são usados para delimitar o comando.

A função +run+ executa o comando:
(((run)))((("função", "Base", "run", see="run"))))

[source,@julia-repl-test chap14]
----
julia> run(cmd);
olá
----

O +olá+ é o resultado do comando echo, mandado para +STDOUT+. A função +run+ retorna um objeto de processo, e gera um +ErrorException+ se o comando externo falha ao executar com sucesso.

Se você quer ler a saída do comando externo, +read+ pode ser usado como alternativa:
(((read)))

[source,@julia-repl-test chap14]
----
julia> a = read(cmd, String)
"olá\n"
----

> Por exemplo, a maioria dos sistemas Unix fornecem um comando chamado +md5sum+ ou +md5+ que lê o conteúdo de um arquivo e calcula a “checksum”. Você pode ler mais sobre MD5 em https://en.wikipedia.org/wiki/Md5. Este comando fornece uma maneira eficiente de verificar se dois arquivos possuem o mesmo conteúdo. A probabilidade de que conteúdos diferentes produzem a mesma checksum é muito pequena (isto é, improvável de acontecer antes que o universo entre em colapso).

Reviewer: Veja se fica bom:
```suggestion
Por exemplo, a maioria dos sistemas Unix fornecem um comando chamado +md5sum+ ou +md5+ que lê o conteúdo de um arquivo e calcula uma “soma de verificação” para checar a integridade dos dados. Você pode ler mais sobre MD5 em https://pt.wikipedia.org/wiki/MD5. Este comando fornece uma maneira eficiente de verificar se dois arquivos possuem o mesmo conteúdo. A probabilidade de que conteúdos diferentes produzem a mesma soma de verificação é muito pequena (isto é, improvável de acontecer antes que o universo entre em colapso).
```

false
----

> Em outras palavras, serialization e em seguida deserialization tem o mesmo efeito de copiar o objeto.

Reviewer: ```suggestion
Em outras palavras, serialização e em seguida desserialização tem o mesmo efeito de copiar o objeto.
```

    contador = 0
    for linha in eachline(nomedoarquivo)
        contador += 1
    end
    contador
end

print(contalinha("wc.jl"))
----

[source,@julia-eval]
----
arquivo = open("wc.jl", "w")
print(arquivo, """function contalinha(nomedoarquivo)
    contador = 0
    for linha in eachline(nomedoarquivo)
        contador += 1
    end
    contador
end

print(contalinha("wc.jl"))""")
close(arquivo)
----

Se você executar esse programa, ele lê a si mesmo e imprime o número de linhas de um arquivo, que é 9. Você também pode incluir ele no REPL assim:
(((include)))((("função", "Base", "include", see="include")))

[source,jl-con]
----
julia> include("wc.jl")
9
----

O Julia introduz módulos para criar uma área de trabalho de variáveis separada, isto é, novos escopos globais.

Um módulo inicia com a palavra-chave +module+ e termina com +end+. Conflitos de nomes são evitados entre suas próprias definições de alto nível e aquelas encontradas em código de outra pessoa. +import+ permite o controle de quais nomes de outros módulos são visíveis e +export+ específica quais nomes são públicos, isto é, podem ser usados fora do módulo sem a necessidade de serem prefixados com o nome do módulo.

> (((escopo)))(((módulo)))((("palavra-chave", "módulo", see="módulo")))(((end)))(((import)))((("palavra-chave", "import", see="import")))(((export)))((("palavra-chave", "export", see="export")))(((contalinha)))((("module", "contalinha", see="contalinha")))((("function", "contalinha", "contalinha", see="contalinha")))

Reviewer 2: ```suggestion
(((escopo)))(((módulo)))((("palavra-chave", "módulo", see="módulo")))(((end)))(((import)))((("palavra-chave", "import", see="import")))(((export)))((("palavra-chave", "export", see="export")))(((ContaLinha)))((("module", "ContaLinha", see="ContaLinha")))((("function", "ContaLinha", "contalinha", see="contalinha")))
```

[source,julia]
----

> module contalinha

Reviewer 2: ```suggestion
module ContaLinha
```

    export contalinha

    function contalinha(nomedoarquivo)
        contador = 0
        for linha in eachline(nomedoarquivo)
            contador += 1
        end
        contador
    end
end
----

> O módulo objeto +contalinha+ fornece +contalinha+:

Reviewer 2: ```suggestion
O objeto de tipo módulo +ContaLinha+ fornece +contalinha+:
```

[source,julia]
----

> julia> using contalinha

Reviewer 2: ```suggestion
julia> using ContaLinha
```

julia> contalinha("wc.jl")
11
----

===== Exercício 14-1

> Digite este exemplo em um arquivo chamado _wc.jl_, inclua-o no REPL e insira +using contalinha+.

Reviewer 2: ```suggestion
Digite este exemplo em um arquivo chamado _wc.jl_, inclua-o no REPL e insira +using ContaLinha+.
```

[WARNING]
====
Se você importar um módulo que já foi importado, o Julia não faz nada. Ele não relê o arquivo, mesmo que ele tenha sido alterado.

Se você quer recarregar o módulo, você tem que reiniciar o REPL. O pacote +Revise+ existe para que você possa manter suas sessões rodando por mais tempo (ver https://github.com/timholy/Revise.jl).
====

> === Depurando

Reviewer: ```suggestion
=== Depuração
```

 4
----

> A função embutida +repr+ ou +dump+ podem ajudar. Elas recebem qualquer objeto como argumento e retornam uma string representando o objeto.

Reviewer: ```suggestion
As funções embutidas +repr+ ou +dump+ podem ajudar. Elas recebem qualquer objeto como argumento e retornam uma string representando o objeto.
```

 4
----

> As funções embutidas +repr+ ou +dump+ podem ajudar. Elas recebem qualquer objeto como argumento e retornam uma string representando o objeto.

Reviewer: ```suggestion
As funções internas +repr+ ou +dump+ podem ajudar. Elas recebem qualquer objeto como argumento e retornam uma string representando o objeto.
```

String "1 2\t 3\n 4"
----

Isso pode ser útil para depurar.
(((depurando)))

Um outro problema que você pode encontrar é que sistemas diferentes usam caracteres diferentes para indicar o final da linha. Alguns sistemas usam uma nova linha, representada por +\n+. Outros usam um caractere de retorno, representado por +\r. Alguns usam ambos. Se você move arquivos entre sistemas diferentes, essas inconsistências podem causar problemas.
(((\r)))

> Para a maioria dos sistemas, existem aplicações que convertem de um formato para o outro. Você pode achá-los (e ler mais a respeito deste problema) em https://en.wikipedia.org/wiki/Newline. Ou, é claro, você pode escrever um por conta própria.

Reviewer: ```suggestion
Para a maioria dos sistemas, existem aplicações que convertem de um formato para o outro. Você pode achá-los (e ler mais a respeito deste problema) em https://pt.wikipedia.org/wiki/Nova_linha. Ou, é claro, você pode escrever um por conta própria.
```

=== Glossário

persistente::
Pertencente a um programa que é executado indefinidamente e mantém pelo menos alguns de seus dados em armazenamento permanente.

> (((persistent)))

Reviewer: ```suggestion
(((persistente)))
```

arquivo de texto::
Uma sequência de caracteres guardados em armazenamento permanente como um disco rígido.
(((arquivo de texto)))

diretório::
Uma coleção de arquivos com nome, também chamada de pasta.
(((diretório)))

> path::

Reviewer: ```suggestion
caminho::
```

Uma string que identifica um arquivo.

> (((path)))

Reviewer: ```suggestion
(((caminho)))
```

> path relativo::

Reviewer: ```suggestion
caminho relativo::
```

> Um path que inicia no diretório atual.

Reviewer: ```suggestion
Um caminho que inicia no diretório atual.
```

> (((path relativo)))

Reviewer: ```suggestion
(((caminho relativo)))
```

> path absoluto::

Reviewer: ```suggestion
caminho absoluto::
```

> Um path que inicia do diretório mais a cima no sistema de arquivos.

Reviewer: ```suggestion
Um caminho que inicia do diretório mais a cima no sistema de arquivos.
```

> (((path absoluto)))

Reviewer: ```suggestion
(((caminho absoluto)))
```

catch::
Prevenir uma exceção de terminar um programa usando as declarações +try pass:[...] catch pass:[...] finally+.
(((catching)))

banco de dados::
Um arquivo cujo conteúdo está organizado como um dicionário com chaves que correspondem a valores.
(((banco de dados)))

shell::
Um programa que permite usuários a digitar comandos e em seguida executá-los iniciando outros programas.
(((shell)))

objeto comando::
Um objeto que representa um comando shell, permitindo um programa em Julia a executar comandos e ler os resultados.
(((objeto comando)))

=== Exercícios

[[ex14-1]]
===== Exercício 14-2

> Escreva uma função chamada +sed+ que recebe como argumento uma string de padrão, uma string de substituição, e dois nomes de arquivos; ela deva ler o primeiro arquivo e escrever o conteúdo no segundo arquivo (criando ele ser for necessário). Se o padrão aparece em qualquer lugar no arquivo, ele deve ser substituido pela string de substituição.

Reviewer: ```suggestion
Escreva uma função chamada +sed+ que recebe como argumento uma string padrão, uma string de substituição e dois nomes de arquivos; ela deva ler o primeiro arquivo e escrever o conteúdo no segundo arquivo (criando ele ser for necessário). Se o padrão aparece em qualquer lugar no arquivo, ele deve ser substituido pela string de substituição.
```

(((sed)))((("função", "definido pelo programador", "sed", see="sed")))

> Se um erro ocorre durante a abertura, leitura, escrita ou fechamento dos arquivos, seu programa deve capturar a exceção, exibir a mensagem de erro, e sair.

Reviewer: ```suggestion
Se um erro ocorre durante a abertura, leitura, escrita ou fechamento dos arquivos, seu programa deve capturar a exceção, exibir a mensagem de erro e sair.
```

[[ex14-2]]
===== Exercício 14-3

> Se você fez <<ex12-2>>, você verá que um dicionário é criado que mapeia uma string ordenada de letras para uma lista de palavras que podem ser escritas com estas letras.Por exemplo, +"opst"+ é mapeado para a lista +["opts", "post", "pots", "spot", "stop", "tops"]+.

Reviewer: ```suggestion
Se você fez <<ex12-2>>, você verá que um dicionário é criado que mapeia uma string ordenada de letras para uma lista de palavras que podem ser escritas com estas letras. Por exemplo, +"opst"+ é mapeado para a lista +["opts", "post", "pots", "spot", "stop", "tops"]+.
```

String "1 2\t 3\n 4"
----

Isso pode ser útil para depurar.

> (((depurando)))

Reviewer: ```suggestion
(((depuração)))
```

> Um outro problema que você pode encontrar é que sistemas diferentes usam caracteres diferentes para indicar o final da linha. Alguns sistemas usam uma nova linha, representada por +\n+. Outros usam um caractere de retorno, representado por +\r. Alguns usam ambos. Se você move arquivos entre sistemas diferentes, essas inconsistências podem causar problemas.

Reviewer 2: ```suggestion
Um outro problema que você pode encontrar é que sistemas diferentes usam caracteres diferentes para indicar o final da linha. Alguns sistemas usam uma nova linha, representada por +\n+. Outros usam um caractere de retorno, representado por +\r+. Alguns usam ambos. Se você mover arquivos entre sistemas diferentes, essas inconsistências podem causar problemas.
```

(((\r)))

Para a maioria dos sistemas, existem aplicações que convertem de um formato para o outro. Você pode achá-los (e ler mais a respeito deste problema) em https://pt.wikipedia.org/wiki/Nova_linha. Ou, é claro, você pode escrever um por conta própria.

=== Glossário

persistente::

> Pertencente a um programa que é executado indefinidamente e mantém pelo menos alguns de seus dados em armazenamento permanente.

Reviewer 2: ```suggestion
Referente a um programa que é executado indefinidamente e mantém pelo menos alguns de seus dados em armazenamento permanente.
```

(((persistente)))

arquivo de texto::
Uma sequência de caracteres guardados em armazenamento permanente como um disco rígido.
(((arquivo de texto)))

diretório::
Uma coleção de arquivos com nome, também chamada de pasta.
(((diretório)))

caminho::
Uma string que identifica um arquivo.
(((caminho)))

caminho relativo::
Um caminho que inicia no diretório atual.
(((caminho relativo)))

caminho absoluto::

> Um caminho que inicia do diretório mais a cima no sistema de arquivos.

Reviewer 2: ```suggestion
Um caminho que inicia do diretório mais acima no sistema de arquivos.
```

(((caminho absoluto)))

catch::
Prevenir uma exceção de terminar um programa usando as declarações +try pass:[...] catch pass:[...] finally+.
(((catching)))

banco de dados::
Um arquivo cujo conteúdo está organizado como um dicionário com chaves que correspondem a valores.
(((banco de dados)))

shell::
Um programa que permite usuários a digitar comandos e em seguida executá-los iniciando outros programas.
(((shell)))

objeto comando::
Um objeto que representa um comando shell, permitindo um programa em Julia a executar comandos e ler os resultados.
(((objeto comando)))

=== Exercícios

[[ex14-1]]
===== Exercício 14-2

> Escreva uma função chamada +sed+ que recebe como argumento uma string padrão, uma string de substituição e dois nomes de arquivos; ela deva ler o primeiro arquivo e escrever o conteúdo no segundo arquivo (criando ele ser for necessário). Se o padrão aparece em qualquer lugar no arquivo, ele deve ser substituido pela string de substituição.

Reviewer 2: ```suggestion
Escreva uma função chamada +sed+ que recebe como argumento uma string padrão, uma string de substituição e dois nomes de arquivos; ela deve ler o primeiro arquivo e escrever o conteúdo no segundo arquivo (o criando se for necessário). Se o padrão aparece em qualquer lugar no arquivo, ele deve ser substituido pela string de substituição.
```

(((sed)))((("função", "definido pelo programador", "sed", see="sed")))

Se um erro ocorre durante a abertura, leitura, escrita ou fechamento dos arquivos, seu programa deve capturar a exceção, exibir a mensagem de erro e sair.

[[ex14-2]]
===== Exercício 14-3

Se você fez <<ex12-2>>, você verá que um dicionário é criado que mapeia uma string ordenada de letras para uma lista de palavras que podem ser escritas com estas letras. Por exemplo, +"opst"+ é mapeado para a lista +["opts", "post", "pots", "spot", "stop", "tops"]+.

Escreva um módulo que importe +anagramsets+ e forneça duas novas funções: +guardaanagramas+ deve guardar o dicionário de anagramas usando +JLD2+; +leranagrams+ deve buscar uma palavra e retornar uma lista de seus anagramas.

> (((guardaanagramas)))((("função", "programmer-defined", "guardaanagramas", see="guardaanagramas")))(((leranagramas)))((("função", "definido pelo programador", "leranagramas", see="leranagramas")))

Reviewer 2: ```suggestion
(((guardaanagramas)))((("função", "definido pelo programador", "guardaanagramas", see="guardaanagramas")))(((leranagramas)))((("função", "definido pelo programador", "leranagramas", see="leranagramas")))
```

[[ex14-3]]
===== Exercício 14-4

Em uma grande coleção de arquivos MP3, podem haver mais de uma cópia da mesma música, guardada em diretórios diferentes ou com nomes de arquivo diferentes. O objetivo deste exercício é procurar por duplicatas.

. Escreva um programa que busca um diretório e todos seus subdiretórios, recursivamente, e retorna uma lista de paths completos para todos os arquivos com um dado sufixo (como _.mp3_).

> . Para reconhecer duplicatas, você pode usar +md5sum+ ou +md5+ para computar a “checksum” de cada arquivo. Se dois arquivos tem a mesma checksum, eles provavelmente tem o mesmo conteúdo.

Reviewer: ```suggestion
. Para reconhecer duplicatas, você pode usar +md5sum+ ou +md5+ para computar uma "verificação de soma” de cada arquivo. Se dois arquivos tem a mesma checksum, eles provavelmente tem o mesmo conteúdo.
```

Author: acho melhor não traduzir, checksum é um termo bastante recorrente. O que acha @abelsiqueira ?

Reviewer 2: Concordo

> . Para verificar novamente, você pode usar o comando Unix +diff+.

Reviewer: Gerar linha "625"

> . Para verificar novamente, você pode usar o comando Unix +diff+.

Reviewer 2: ```suggestion
. Para verificar novamente, você pode usar o comando Unix +diff+.

```
